File: leetcode/range_sum.cpp
#include <vector>

// number: 303
// section: assessments
// difficulty: easy
// tags: array, design, prefix sum, meta

// constraints
// 1 <= nums.length <= 10^4
// -10^5 <= nums[i] <= 10^5
// 0 <= left <= right < nums.length
// At most 10^4 calls will be made to sum_range.

// solution: prefix sum
// complexity
// run-time: O(n)
// space: O(n)
class NumArray {
public:
    NumArray(const vector<int>& nums)
    : d_nums(nums)
    {
        calcPrefixSum();
    }

    int sumRange(int left, int right) {
        return d_prefixSum[right]-d_prefixSum[left]+d_nums[left];
    }

private:
    void calcPrefixSum() {
        if (d_nums.empty()) {
            return;
        }

        d_prefixSum.push_back(d_nums[0]);

        for (size_t i = 1; i != d_nums.size(); ++i) {
            d_prefixSum.push_back(d_prefixSum.back() + d_nums[i]);
        }
    }

    std::vector<int> d_nums;
    std::vector<int> d_prefixSum;
};

// Your NumArray object will be instantiated and called as such:
NumArray* obj = new NumArray(nums);
int param_1 = obj->sumRange(left,right);

File: leetcode/range_sum.py
#!/usr/bin/env python3

import sys
import unittest

# number: 303
# title: Range Sum Query - Immutable
# url: https://leetcode.com/problems/range-sum-query-immutable/
# section: assessments
# difficulty: easy
# tags: array, design, prefix sum, meta

# constraints
# 1 <= nums.length <= 10^4
# -10^5 <= nums[i] <= 10^5
# 0 <= left <= right < nums.length
# At most 10^4 calls will be made to sum_range.

# solution: prefix sum
# complexity
# run-time: O(n)
# space: O(n)
class NumArray:
    def __init__(self, nums):
        self.__nums = nums
        self.__prefix_sums = []
        self.calc_prefix_sum(nums)

    def sum_range(self, left: int, right: int) -> int:
        return self.__prefix_sums[right]-self.__prefix_sums[left]+self.__nums[left]

    def calc_prefix_sum(self, nums):
        if not nums:
            return

        self.__prefix_sums.append(nums[0])

        for i in range(1, len(nums)):
            self.__prefix_sums.append(self.__prefix_sums[-1] + nums[i])

class TestRangeSum(unittest.TestCase):
    def test1(self):
        num_array = NumArray([-2, 0, 3, -5, 2, -1])
        self.assertEqual(num_array.sum_range(0, 2), 1)
        self.assertEqual(num_array.sum_range(2, 5), -1)
        self.assertEqual(num_array.sum_range(0, 5), -3)

if __name__ == '__main__':
    sys.exit(unittest.main())
